add details on method call syntax in oop chapter (#100)
flaviut/nim-by-example: content/content/oop.md

  
var sparky = Animal(name: "Sparky", age: 10)

sparky.speak("Hi")
# this is the same as the below
speak(sparky, "Hi")
```

As you might have guessed, the method call syntax is not restricted to object types.

``` nimrod
proc double(num: int):int =
  return num*2
  
double(10)
# same as
10.double()
# you can even drop the empty paranthesis
10.double
```

Now, when using UFCS to think about `proc`s as member functions, there is an important point to keep in mind about mutability. Consider this example.

``` nimrod
type Animal = object
  name: string
  age: int
  
proc incAge(self: Animal) =
  self.age += 1
  
var sparky = Animal(name: "Sparky", age: 3)
```

This fails compiling with the following error:
``` console
Error: type mismatch: got <int, int literal(1)>
but expected one of:
proc `+=`[T: SomeInteger](x: var T; y: T)
  first type mismatch at position: 1
  required type for x: var T: SomeInteger
  but expression 'self.age' is immutable, not 'var'
proc `+=`[T: float | float32 | float64](x: var T; y: T)
  first type mismatch at position: 1
  required type for x: var T: float or float32 or float64
  but expression 'self.age' is of type: int

expression: self.age += 1

> ```

Reviewer: I would remove these--I don't think the more complex error shows anything that the simpler one doesn't.

Author: An even more inscrutable (to a beginner like me, at least) error message (https://github.com/nim-lang/Nim/issues/20779) was what got me to writing this section.

May be, I can remove this example and instead just have a note on function type mismatch errors in these scenarios.

Reviewer: Ok, I see where you're coming from. I think it's reasonable to keep--but that the simple example should come first to familiarize the reader with the situation. I think it'd also be a good idea to remove the second half of the error message, about the "float"s.

If you try something simpler:
``` nimrod
proc setName(self: Animal, name: string) =
  self.name = name
```
the error message is a lot clearer:
``` console
Error: 'self.name' cannot be assigned to
```

> The reason for this is, that arguments to `proc`s are immutable. This is simply solved by marking mutable arguments using `var` as shown below:

Reviewer: ```suggestion
This happens because arguments to `proc` are immutable by default. The error can be fixed by marking the argument as mutable using `var`:
```

``` nimrod
proc incAge(self: var Animal) =
  self.age += 1

proc setName(self: var Animal, name: string) =
  self.name = name
```

Now, everything works.

> Note that this problem does not occur with `ref` objects, so the following works correctly:

Reviewer: ```suggestion
Note that `ref` objects can be mutated, and the following works correctly:
```

``` nimrod
type Animal = ref object
  name: string
  age: int
  
proc incAge(self: Animal) =
  self.age += 1
```

Why is this? This is because self is just a pointer and is not being mutated. If you try to change the self variable itself, it would not work.

## Inheritance

> Nim supports inheritance and dynamically dispatched methods, which are virtual procedures. Methods work in the same way as [procs](/procs/), including UFCS, but the runtime type of the object that a method is called with is used to determine which version to call.

Reviewer: ```suggestion
Nim supports inheritance and dynamically dispatched (also known as "virtual") methods. Methods work in the same way as [procs](/procs/), including UFCS, but the runtime type of the object that a method is called with is used to determine which version to call.
```
